runtime-modules/working-team/src/checks.rs

```diff
@@ -0,0 +1,160 @@
+use crate::{Instance, JobOpening, JobOpeningType, TeamWorkerId, Trait};
+
+use super::Error;
+use frame_support::dispatch::{DispatchError, DispatchResult};
+use frame_support::{ensure, StorageMap, StorageValue};
+use sp_std::collections::btree_set::BTreeSet;
+use system::{ensure_root, ensure_signed};
+
+use crate::types::ApplicationInfo;
+
+// Check opening: verifies opening description length.
+pub(crate) fn ensure_opening_description_is_valid<T: Trait<I>, I: Instance>(
+    text: &[u8],
+) -> DispatchResult {
+    <crate::OpeningDescriptionTextLimit<I>>::get()
+        .ensure_valid(
+            text.len(),
+            Error::<T, I>::OpeningDescriptionTooShort.into(),
+            Error::<T, I>::OpeningDescriptionTooLong.into(),
+        )
+        .map_err(|e| DispatchError::Other(e))
+}
+
+// Check opening: verifies origin and opening type compatibility.
+pub(crate) fn ensure_origin_for_opening_type<T: Trait<I>, I: Instance>(
+    origin: T::Origin,
+    opening_type: JobOpeningType,
+) -> DispatchResult {
+    match opening_type {
+        JobOpeningType::Regular => {
+            // Ensure lead is set and is origin signer.
+            ensure_origin_is_active_leader::<T, I>(origin)
+        }
+        JobOpeningType::Leader => {
+            // Council proposal.
+            ensure_root(origin).map_err(|err| err.into())
+        }
+    }
+}
+
+// Check opening: returns the opening by id if it is exists.
+pub(crate) fn ensure_opening_exists<T: Trait<I>, I: Instance>(
+    opening_id: &T::OpeningId,
+) -> Result<JobOpening<T::BlockNumber, T::ApplicationId>, Error<T, I>> {
+    ensure!(
+        <crate::OpeningById::<T, I>>::contains_key(opening_id),
+        Error::<T, I>::OpeningDoesNotExist
+    );
+
+    let opening = <crate::OpeningById<T, I>>::get(opening_id);
+
+    Ok(opening)
+}
+
+// Check application: verifies application description length.
+pub(crate) fn ensure_application_description_is_valid<T: Trait<I>, I: Instance>(
+    text: &[u8],
+) -> DispatchResult {
+    <crate::ApplicationDescriptionTextLimit<I>>::get()
+        .ensure_valid(
+            text.len(),
+            Error::<T, I>::JobApplicationDescriptionTooShort.into(),
+            Error::<T, I>::JobApplicationDescriptionTooLong.into(),
+        )
+        .map_err(|e| DispatchError::Other(e))
+}
+
+// Check application: ensures that nobody applies for the same opening twice.
+pub(crate) fn ensure_member_has_no_active_application_on_opening<T: Trait<I>, I: Instance>(
```

Comment:
We are no longer rationing access to apply based on the application stake. If we still insist on having a limit, then an opening can be attacked by having someone just fill up all spots, excluding anyone else. The solution is that we should have no limit, nor do we actually need one for any reason. This means all applicants live in a separate map, not as elements in an inline set in the opening. If we want to maintain this rule of applying only once per member to each opening, then a double map `OpeningIdxMemberId -> Application` could be sufficient, but I am not sure its even worth enforcing this rule. In some cases, you cuold even want to ahve one member having multiple worker roles in one group, so why even bother with trying to prevent applying more than once in one opening?

However we store the applications, they are cleaned out of storage as people unstake the role stake when they lost, or if they are winners. The only edge case is if there is no stake for the role, in that case it would live in storage forever. To avoid this would could hard code a lower bound > tx fee of cleanup.

We no longer need to iterate all applications, as we are not managing their stake. This means we no longer should attempt to limit the number of applicants. This means the number of applicants should be unconstrained, i.e.

Reply:
Removed.